pkg/controller/pravegacluster/upgrade.go

```diff
@@ -315,6 +315,9 @@ func (r *ReconcilePravegaCluster) checkUpdatedPods(sts *appsv1.StatefulSet, vers
 
 		if !util.IsPodReady(pod) {
 			// At least one updated pod is still not ready
+			if pod.Status.ContainerStatuses[0].State.Waiting.Reason == "ImagePullBackOff" {
```

Comment:
This solves only the "ImagePullBackOff" reason for update failure. There could be more reasons why an update could fail. Ideally we want to capture any unrecoverable upgrade failure and report it.
What can we do to solve this more generically?

Please make sure this is tested for positive and negative test case scenarios.
Please add unit tests to cover these.

Reply:
In case upgrade fails due to a `ContainerState` terminating with a non-zero exit status, Kubernetes tries to bring up the pod again by restarting the pod. After a certain number of retries it detects the failure and the process terminates with the message `"error syncing cluster version, need manual intervention. failed to sync bookkeeper version. pod bar-pravega-bookie-0 is restarting"`

The `Error` Status is also appropriately updated, as can be seen from the following output

**kubectl describe PravegaCluster**
```
Status:
  Conditions:
    Last Transition Time:  2019-08-19T06:02:15Z
    Last Update Time:      2019-08-19T06:02:15Z
    Status:                False
    Type:                  Upgrading
    Last Transition Time:  2019-08-19T06:01:43Z
    Last Update Time:      2019-08-19T06:01:43Z
    Status:                False
    Type:                  PodsReady
    Last Transition Time:  2019-08-19T06:02:15Z
    Last Update Time:      2019-08-19T06:02:15Z
    Message:               failed to sync bookkeeper version. pod bar-pravega-bookie-0 is restarting
    Reason:                UpgradeFailed
    Status:                True
    Type:                  Error
  Current Replicas:        5
  Current Version:         0.5.0-2222.8ba5431
```